app/services/folio/locations.rb

```diff
@@ -7,12 +7,75 @@ module Folio
   class Locations
     include Singleton
 
+    HASH_FIELDS = %w[details metadata].freeze
+
     # @param [String] code the Folio location code
     # @return [String] the Folio location label
     def self.label(code:)
       instance.data.dig(code, 'discoveryDisplayName')
     end
 
+    # Output a CSV of all locations, flattening details and converting UUIDs to codes
+    # @return [String] comma-separated data for all locations
+    # rubocop:disable Metrics/MethodLength
+    def self.to_csv
+      headers = []
+      detail_headers = []
+
+      csv_data = instance.data.each_value.map do |loc|
+        # don't mutate original data
+        location = loc.dup
+
+        # flatten detail fields
+        if location['details'].present?
+          location['details'].each do |key, val|
+            next if val.blank?
+
+            location["details.#{key}"] = val
+            detail_headers |= ["details.#{key}"]
+          end
+        end
+
+        # look up UUIDs and convert to codes
+        location['institutionCode'] = Folio::Types.institutions.dig(location['institutionId'], 'code')
+        location['campusCode'] = Folio::Types.campuses.dig(location['campusId'], 'code')
+        location['libraryCode'] = Folio::Types.libraries.dig(location['libraryId'], 'code')
+        location['primaryServicePointCode'] = Folio::Types.service_points.dig(location['primaryServicePoint'], 'code')
+        location['servicePointCodes'] = location['servicePointIds']&.map { |id| Folio::Types.service_points.dig(id, 'code') }
+
+        # remove fields that are now unused
+        location.each { |key, _val| key.match?(/Ids?$/) && location.delete(key) }
+        location.delete('primaryServicePoint') # is a UUID, despite how FOLIO named it
+        location.delete('servicePoints') # always empty
+        location.delete('metadata')
+        location.delete('details')
+
+        # save headers for CSV
+        headers |= (location.keys - detail_headers)
+        location
+      end
+
+      # change the order of the headers so that variable-length object keys are at the end
+      headers |= %w[institutionCode campusCode libraryCode primaryServicePointCode servicePointCodes]
```

Comment:
Sorry if I was unclear. I think having `institutionCode campusCode libraryCode` and the location's `code` at the beginning is a very good thing. It's just the service point data I thought we should relegate.

Reply:
Reordered